lib/src/message_pack/encoder.dart

```diff
@@ -0,0 +1,294 @@
+    builder.addByte(MessagePackFormat.fixArrayMin | length);
+  } else if (length <= 0xFFFF) {
+    builder.addByte(MessagePackFormat.array16);
+    builder.add(_uint16ToBytes(length));
+  } else if (length <= 0xFFFFFFFF) {
+    builder.addByte(MessagePackFormat.array32);
+    builder.add(_uint32ToBytes(length));
+  } else {
+    throw ArgumentError(
+      'Array length exceeds MessagePack limit (2^32-1 elements): $length',
+    );
+  }
+  for (final item in value) {
+    encodeMsgPackValue(item, builder);
+  }
+}
+
+/// Encode array header
+void encodeArrayHeader(int length, BytesBuilder builder) {
+  if (length <= 15) {
+    builder.add(Uint8List.fromList([MessagePackFormat.fixArrayMin | length]));
+  } else if (length <= 0xFFFF) {
+    final lengthBytes = _uint16ToBytes(length);
+    builder.add(
+      Uint8List.fromList([MessagePackFormat.array16, ...lengthBytes]),
+    );
+  } else {
+    final lengthBytes = _uint32ToBytes(length);
+    builder.add(
+      Uint8List.fromList([MessagePackFormat.array32, ...lengthBytes]),
+    );
+  }
+}
+
+/// Encode map header
+void encodeMapHeader(int length, BytesBuilder builder) {
+  if (length <= 15) {
+    builder.add(Uint8List.fromList([MessagePackFormat.fixMapMin | length]));
+  } else if (length <= 0xFFFF) {
+    final lengthBytes = _uint16ToBytes(length);
+    builder.add(Uint8List.fromList([MessagePackFormat.map16, ...lengthBytes]));
+  } else {
+    final lengthBytes = _uint32ToBytes(length);
+    builder.add(Uint8List.fromList([MessagePackFormat.map32, ...lengthBytes]));
+  }
+}
+
+/// Encodes a Map, selecting the most compact format.
+void _encodeMap(Map value, BytesBuilder builder) {
+  final length = value.length;
+
+  if (length <= 15) {
+    builder.addByte(MessagePackFormat.fixMapMin | length);
+  } else if (length <= 0xFFFF) {
+    builder.addByte(MessagePackFormat.map16);
+    builder.add(_uint16ToBytes(length));
+  } else if (length <= 0xFFFFFFFF) {
+    builder.addByte(MessagePackFormat.map32);
+    builder.add(_uint32ToBytes(length));
+  } else {
+    throw ArgumentError(
+      'Map length exceeds MessagePack limit (2^32-1 elements): $length',
+    );
+  }
+  value.forEach((key, val) {
+    encodeMsgPackValue(key, builder);
+    encodeMsgPackValue(val, builder);
+  });
+}
+
+/// Encodes an Extension type, selecting the most compact format.
+void _encodeExtension(MessagePackExt ext, BytesBuilder builder) {
+  final length = ext.data.length;
+
+  builder.addByte(switch (length) {
+    1 => MessagePackFormat.fixext1,
+    2 => MessagePackFormat.fixext2,
+    4 => MessagePackFormat.fixext4,
+    8 => MessagePackFormat.fixext8,
+    16 => MessagePackFormat.fixext16,
+    <= 0xFF => MessagePackFormat.ext8,
+    <= 0xFFFF => MessagePackFormat.ext16,
+    <= 0xFFFFFFFF => MessagePackFormat.ext32,
+    _ => throw ArgumentError(
+      'Extension data length exceeds MessagePack limit (2^32-1 bytes): $length',
+    ),
+  });
+
+  if (![1, 2, 4, 8, 16].contains(length)) {
```

Comment:
Either everything remains in a single if/else block or we should consider simplifying line `224-24` to run within a single switch. E.g 

```dart
  builder.add(switch (length) {
    1 => [MessagePackFormat.fixext1],
    2 => [MessagePackFormat.fixext2],
    4 => [MessagePackFormat.fixext4],
    8 => [MessagePackFormat.fixext8],
    16 => [MessagePackFormat.fixext16],
    <= 0xFF => [MessagePackFormat.ext8, length],
    <= 0xFFFF => [MessagePackFormat.ext16, ..._uint16ToBytes(length)],
    <= 0xFFFFFFFF => [MessagePackFormat.ext32, ..._uint32ToBytes(length)],
    _ => throw ArgumentError(
      'Extension data length exceeds MessagePack limit (2^32-1 bytes): $length',
    ),
  });